Zend/zend_operators.h

```diff
@@ -919,6 +919,98 @@ static zend_always_inline void zend_unwrap_reference(zval *op) /* {{{ */
 }
 /* }}} */
 
+static zend_always_inline const char *zend_strnichr(const char *ptr, char value, size_t num)
+{
+	char value_lower = zend_tolower_ascii(value);
+	char value_upper = zend_toupper_ascii(value);
+
+	if (value_lower == value_upper) {
+		return (const char *)memchr(ptr, value, num);
+	}
+
+	const char *ptr_lower = (const char *)memchr(ptr, value_lower, num);
+	const char *ptr_upper = (const char *)memchr(ptr, value_upper, num);
+	if (ptr_lower && ptr_upper) {
+		return ptr_lower < ptr_upper ? ptr_lower : ptr_upper;
+	}
+	return ptr_lower ? ptr_lower : ptr_upper;
+}
+
+static zend_always_inline bool zend_strnieq(const char *ptr1, const char *ptr2, size_t num)
+{
+	for (size_t i = 0; i < num; i++) {
+		if (zend_tolower_ascii(*ptr1++) != zend_tolower_ascii(*ptr2++)) {
+			return 0;
+		}
+	}
+	return 1;
+}
+
+static zend_always_inline const char *
+zend_memnistr(const char *haystack, const char *needle, size_t needle_len, const char *end)
+{
+	const char *p = haystack;
+	size_t off_s;
+
+	ZEND_ASSERT(end >= p);
+
+	if (needle_len == 1) {
+		return (const char *)zend_strnichr(p, *needle, (end-p));
+	} else if (UNEXPECTED(needle_len == 0)) {
+		return p;
+	}
+
+	off_s = (size_t)(end - p);
+
+	if (needle_len > off_s) {
+		return NULL;
+	}
+
+	const char ns_upper = zend_tolower_ascii(*needle);
+	const char ns_lower = zend_toupper_ascii(*needle);
+	const char ne = zend_tolower_ascii(needle[needle_len-1]);
+	end -= needle_len;
+
+	if (ns_upper == ns_lower) {
+		while (p <= end) {
+			if ((p = (const char *)memchr(p, ns_lower, (end-p+1)))) {
+				if (ne == zend_tolower_ascii(p[needle_len-1]) && zend_strnieq(needle+1, p+1, needle_len-2)) {
+					return p;
+				}
+			} else {
+				return NULL;
+			}
+			p++;
+		}
+	} else {
+		const char *next_lower = (const char *)memchr(p, ns_lower, (end-p+1));
+		const char *next_upper = (const char *)memchr(p, ns_upper, (end-p+1));
+		while (p <= end) {
+			if (next_lower && (!next_upper || next_lower < next_upper)) {
+				p = next_lower;
+				next_lower = NULL;
+			} else if (next_upper) {
+				p = next_upper;
+				next_upper = NULL;
+			} else {
+				return NULL;
+			}
+			if (ne == zend_tolower_ascii(p[needle_len-1]) && zend_strnieq(needle+1, p+1, needle_len-2)) {
+				return p;
+			}
+			p++;
+			if (!next_lower) {
+				next_lower = (const char *)memchr(p, ns_lower, (end-p+1));
+			}
+			if (!next_upper) {
+				next_upper = (const char *)memchr(p, ns_upper, (end-p+1));
+			}
```

Comment:
If lower or upper case letter is not present anymore in the remaining haystack both searches are performed anyway with every subsequent iteration. This is where performance dip is coming from.

Reply:
I missed that, thanks!